src/thor/costmatrix.cc

```diff
@@ -260,559 +238,528 @@ void CostMatrix::SourceToTarget(Api& request,
+  }
 
-      // Skip this edge if permanently labeled (best path already found to this
-      // directed edge) or if no access for this mode.
-      if (es->set() == EdgeSet::kPermanent || !(directededge->forwardaccess() & access_mode_)) {
-        continue;
-      }
+  // keep track of shortcuts
+  uint32_t shortcuts = 0;
+  // If we encounter a node with an access restriction like a barrier we allow a uturn
+  if (!costing_->Allowed(nodeinfo)) {
+    const DirectedEdge* opp_edge = nullptr;
+    const GraphId opp_edge_id = graphreader.GetOpposingEdgeId(pred.edgeid(), opp_edge, tile);
+    // Mark the predecessor as a deadend to be consistent with how the
+    // edgelabels are set when an *actual* deadend (i.e. some dangling OSM geometry)
+    // is labelled
+    pred.set_deadend(true);
+    // Check if edge is null before using it (can happen with regional data sets)
+    return opp_edge && ExpandInner<expansion_direction>(graphreader, index, pred, opp_pred_edge,
+                                                        nodeinfo, pred_idx,
+                                                        {opp_edge, opp_edge_id,
+                                                         edgestatus.GetPtr(opp_edge_id, tile)},
+                                                        shortcuts, tile, offset_time);
+  }
 
-      // Skip this edge if no access is allowed (based on costing method)
-      // or if a complex restriction prevents transition onto this edge.
-      uint8_t restriction_idx = -1;
-      if (!costing_->Allowed(directededge, false, pred, tile, edgeid, offset_time.local_time,
-                             nodeinfo->timezone(), restriction_idx) ||
-          costing_->Restricted(directededge, pred, edgelabels, tile, edgeid, true, nullptr,
-                               offset_time.local_time, nodeinfo->timezone())) {
-        continue;
-      }
+  // catch u-turn attempts
+  bool disable_uturn = false;
+  EdgeMetadata meta = EdgeMetadata::make(node, nodeinfo, tile, edgestatus);
+  EdgeMetadata uturn_meta{};
+
+  // Expand from end node in <expansion_direction> direction.
+  for (uint32_t i = 0; i < nodeinfo->edge_count(); ++i, ++meta) {
+
+    // Begin by checking if this is the opposing edge to pred.
+    // If so, it means we are attempting a u-turn. In that case, lets wait with evaluating
+    // this edge until last. If any other edges were emplaced, it means we should not
+    // even try to evaluate a u-turn since u-turns should only happen for deadends
+    const bool is_uturn = pred.opp_local_idx() == meta.edge->localedgeidx();
+    uturn_meta = is_uturn ? meta : uturn_meta;
+
+    // Expand but only if this isnt the uturn, we'll try that later if nothing else works out
+    disable_uturn =
+        (!is_uturn &&
+         ExpandInner<expansion_direction>(graphreader, index, pred, opp_pred_edge, nodeinfo, pred_idx,
+                                          meta, shortcuts, tile, offset_time)) ||
+        disable_uturn;
+  }
```

Comment:
this is whole motivation for this PR: properly handle uturns. should look familiar :)